### PR TITLE
Support separators in floating point literals

### PR DESCRIPTION
**Pull request by pulpdrew**
Completes https://github.com/AssemblyScript/assemblyscript/issues/1059

- [x] I've read the contributing guidelines

### CODE REVIEW

File: src/tokenizer.ts [hunk 1]
```diff
@@ -1485,12 +1480,45 @@ export class Tokenizer extends DiagnosticEmitter {
...
-        while (this.pos < end && isDecimalDigit(text.charCodeAt(this.pos))) {
-          ++this.pos;
+        this.readDecimalFloatPartial();
+      }
+    }
+    let floatString = text.substring(start, this.pos).replace(/_/g, '');
+    return parseFloat(floatString);
+  }
+
+  private readDecimalFloatPartial() {
```
Comment (MaxGraey) [inline src/tokenizer.ts:1490]:
```suggestion
  private readDecimalFloatPartial(): void {
```

File: src/tokenizer.ts [hunk 2]
```diff
@@ -1485,12 +1480,45 @@ export class Tokenizer extends DiagnosticEmitter {
         ) {
           ++this.pos;
         }
-        while (this.pos < end && isDecimalDigit(text.charCodeAt(this.pos))) {
-          ++this.pos;
+        this.readDecimalFloatPartial();
+      }
+    }
+    let floatString = text.substring(start, this.pos).replace(/_/g, '');
```
Comment (MaxGraey) [inline src/tokenizer.ts:1486]:
```suggestion
    let floatString = text.substring(start, this.pos).replaceAll("_", "");
```

Comment (pulpdrew) [reply src/tokenizer.ts]: This seems to cause an error `TS2339: Property 'replaceAll' does not exist on type 'string'.`

File: tests/parser/numeric-separators.ts [hunk 3]
```diff
@@ -16,3 +19,15 @@
...
+1000._1234; // 6188
+1000.1234_; // 6188
+
+10__00.1234; // 6189
+1000.12__34; // 6189
+
+1_e2; // 6188
+1e_2; // 6188
+1e2_; // 6188
+1e-1__0; // 6189
```
Comment (MaxGraey) [inline tests/parser/numeric-separators.ts:33]: nice! Could you also add this invalid cases:
```ts
_1.0
0_0.0
0_0.0_0
```
and valid case:
```ts
0.0_0
```

File: src/tokenizer.ts (hunk 2 above)
Comment (MaxGraey) [reply src/tokenizer.ts]: just ignore this via `@ts-ignore`

Comment (MaxGraey) [reply src/tokenizer.ts]: btw it should be [exists in `esnext`](https://github.com/microsoft/TypeScript/pull/37476)

Comment (MaxGraey) [reply src/tokenizer.ts]: So I suggest change to `esnext` instead `es2016` [here](https://github.com/AssemblyScript/assemblyscript/blob/master/src/tsconfig.json#L7) instead using `// @ts-ignore`

File: tests/parser/numeric-separators.ts (hunk 3 above)
Comment (pulpdrew) [reply tests/parser/numeric-separators.ts:33]: I am a little confused about some of these error cases.

How would you recommend testing `_1.0`? It seems like it wouldn't be parsed as a float - it would be an identifier `_1` and a float `0`. Should I expect that as output?

All of these seem to be accepted by the typescript compiler and interpreted as decimal floats:
```
0_0.0
0_0.0_0
0_0e0_0
```
Could you point me towards the spec that AssemblyScript is following for this feature? Or could you summarize the specific problem with these examples? Is it just that `0_` is an invalid prefix of any of the fragments?

Thanks!

Comment (MaxGraey) [reply tests/parser/numeric-separators.ts:33]: Hmm, really TS accept this. However by spec leading `0_` should fail: https://github.com/tc39/test262/blob/master/test/language/literals/numeric/numeric-separators/numeric-separator-literal-nzd-nsl-dds-leading-zero-err.js

File: src/tokenizer.ts (hunk 2 above)
Comment (pulpdrew) [reply src/tokenizer.ts]: `// @ts-ignore` worked, but `esnext` did not.

File: tests/parser/numeric-separators.ts (hunk 3 above)
Comment (MaxGraey) [reply tests/parser/numeric-separators.ts:33]: > How would you recommend testing _1.0

Yeah, it should start parsing as identifier, so I removed this from suggestions

File: src/tsconfig.json [hunk 4]
```diff
@@ -4,7 +4,7 @@
     "outDir": "../out",
     "allowJs": false,
     "sourceMap": true,
-    "target": "ES2016",
+    "target": "esnext",
```
Comment (MaxGraey) [inline src/tsconfig.json:7]: it seems also need add `libs: ["esnext", "esnext.string"]`. But better try locally first via `npm run check` and `npm run build`. If it still fail just revert `@ts-ignore`

Comment (pulpdrew) [reply src/tsconfig.json:7]: That worked, thanks!

File: src/tokenizer.ts [hunk 5]
```diff
@@ -1485,12 +1480,50 @@ export class Tokenizer extends DiagnosticEmitter {
...
-        while (this.pos < end && isDecimalDigit(text.charCodeAt(this.pos))) {
-          ++this.pos;
+        this.readDecimalFloatPartial();
+      }
+    }
+    let floatString = text.substring(start, this.pos).replaceAll("_", "");
+    return parseFloat(floatString);
+  }
+
+  private readDecimalFloatPartial(allowLeadingZeroSep: boolean = true): void {
```
Comment (dcodeIO) [inline src/tokenizer.ts:1490]:
```suggestion
  private readDecimalFloatPartial(allowLeadingZeroSep: bool = true): void {
```

File: src/tsconfig.json [hunk 6]
```diff
@@ -4,8 +4,9 @@
     "outDir": "../out",
     "allowJs": false,
     "sourceMap": true,
-    "target": "ES2016",
-    "strict": true
+    "target": "esnext",
+    "strict": true,
+    "lib": ["esnext", "esnext.string"]
```
Comment (dcodeIO) [inline src/tsconfig.json:9]: Since this is inheriting from `std/portable.json`, wdyt of making this change there so all portable AS projects can use `replaceAll`, not just the compiler?

File: src/tokenizer.ts [hunk 7]
```diff
@@ -1485,12 +1480,50 @@ export class Tokenizer extends DiagnosticEmitter {
         ) {
           ++this.pos;
         }
-        while (this.pos < end && isDecimalDigit(text.charCodeAt(this.pos))) {
-          ++this.pos;
+        this.readDecimalFloatPartial();
+      }
+    }
+    let floatString = text.substring(start, this.pos).replaceAll("_", "");
+    return parseFloat(floatString);
```
Comment (dcodeIO) [inline src/tokenizer.ts:1487]:
```suggestion
    return parseFloat(text.substring(start, this.pos).replaceAll("_", ""));
```

File: src/tokenizer.ts [hunk 8]
```diff
@@ -1485,12 +1480,50 @@ export class Tokenizer extends DiagnosticEmitter {
         ) {
           ++this.pos;
         }
-        while (this.pos < end && isDecimalDigit(text.charCodeAt(this.pos))) {
-          ++this.pos;
+        this.readDecimalFloatPartial();
+      }
+    }
+    let floatString = text.substring(start, this.pos).replaceAll("_", "");
```
Comment (MaxGraey) [inline src/tokenizer.ts:1486]: Also could you add `hasSep` boolean variable which set to true if at least one separator found and do `replaceAll` only for this case like:
```ts
  ...
  let result = text.substring(start, this.pos)
  if (hasSep) result = result.replaceAll("_", "");
  return parseFloat(result);
}
```

File: src/tokenizer.ts (hunk 7 above)
Comment (MaxGraey) [reply src/tokenizer.ts:1487]: Perhaps [this](https://github.com/AssemblyScript/assemblyscript/pull/1385/files#r453169401) even better

File: src/tokenizer.ts [hunk 9]
```diff
@@ -1494,37 +1494,78 @@ export class Tokenizer extends DiagnosticEmitter {
   }
 
   readDecimalFloat(): f64 {
-    // TODO: numeric separators (parseFloat can't handle these)
     var text = this.source.text;
-    var pos = this.pos;
```
Comment (pulpdrew) [inline src/tokenizer.ts:1499]: This local variable `pos` was added in #1376 , but made using a helper function awkward. I could either:
- leave things how I have them now
- use a local pos and do `this.pos = pos` before each call to the helper and `pos = this.pos` after each call to the helper
- eliminate the helper function and just inline it all in this function.
- pass a `start` argument into the helper (but that still leaves the problem of returning the end position, since I am already using the return value for indicating whether a separator is present). 

Thoughts?

Comment (MaxGraey) [reply src/tokenizer.ts:1499]: Yeah, that's caching `this.pos` is made some thing little complicated but it's speedup self-bootstrap version of AS. I suggest leave all as is because that cacheing most significant only for `unsafeNext`. Rest methods might leave as is for simplicity

File: src/tokenizer.ts [hunk 10]
```diff
@@ -1494,37 +1494,78 @@ export class Tokenizer extends DiagnosticEmitter {
...
-    }
-    if (pos < end && text.charCodeAt(pos) == CharCode.DOT) {
-      ++pos;
-      while (pos < end && isDecimalDigit(text.charCodeAt(pos))) {
-        ++pos;
-      }
+    var start = this.pos;
+    var hasSep = false;
+
+    hasSep = this.readDecimalFloatPartial(false) || hasSep;
```
Comment (MaxGraey) [inline src/tokenizer.ts:1502]: Having looked again, I think that it makes sense to use an integer counter instead of a boolean value:
```ts
var sepCount = 0;

sepCount += this.readDecimalFloatPartial(false);
...

if (sepCount) result = result.replaceAll("_", "");
return parseFloat(result);


```

**Comment by dcodeIO** [comment]
Thank you! :)
